private-identity-audit/src/lib.rs

```diff
@@ -25,22 +37,98 @@ use uuid::Uuid;
 /// element of a set, PUIS can guess that element with probability 1/the_size_of_the_padded_set.
 pub const SET_SIZE_ANONYMITY_PARAM: usize = 100_000;
 
+pub struct InitialProver;
+pub struct FinalProver;
+
+pub struct VerifierSetGenerator;
+pub struct Verifier;
+
 /// The initial private set of PUIS.
-pub type PrivateUids = Vec<Scalar>;
+pub struct PrivateUids(pub Vec<Scalar>);
+
+impl Encode for PrivateUids {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        SCALAR_SIZE * self.0.len()
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let uuids: Vec<[u8; SCALAR_SIZE]> = self.0.iter().map(|u| u.to_bytes()).collect();
+
+        uuids.encode_to(dest);
+    }
+}
+
+impl Decode for PrivateUids {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let uuids = <Vec<[u8; SCALAR_SIZE]>>::decode(input)?;
+
+        let uuids: Vec<Scalar> = uuids.into_iter().map(Scalar::from_bits).collect();
+
+        Ok(PrivateUids(uuids))
+    }
+}
 
 /// The committed and padded version of the private set of PUIS.
-pub type CommittedUids = Vec<RistrettoPoint>;
+#[derive(Clone, Debug, PartialEq)]
+pub struct CommittedUids(pub Vec<RistrettoPoint>);
+
+impl Encode for CommittedUids {
+    #[inline]
+    fn size_hint(&self) -> usize {
+        RISTRETTO_POINT_SIZE * self.0.len()
+    }
+
+    fn encode_to<W: Output>(&self, dest: &mut W) {
+        let committed_uuids: Vec<[u8; RISTRETTO_POINT_SIZE]> =
+            self.0.iter().map(|u| u.compress().to_bytes()).collect();
+
+        committed_uuids.encode_to(dest);
+    }
+}
+
+impl Decode for CommittedUids {
+    fn decode<I: Input>(input: &mut I) -> Result<Self, CodecError> {
+        let committed_uuids = <Vec<[u8; RISTRETTO_POINT_SIZE]>>::decode(input)?;
+        let committed_uuids: Vec<RistrettoPoint> = committed_uuids
+            .into_iter()
+            .map(|u| {
+                CompressedRistretto(u)
+                    .decompress()
+                    .unwrap_or_else(RistrettoPoint::default)
+            })
+            .collect();
```

Comment:
One of most lovely feature of `.collect()` could help here. It can transform from `Collection< Result<T,E>>` into `Result<Collection<T>, E>`, returning the first error only (which is valid for this case).
I don't remember if it does an early return on first error case, which could be an "extra" here.

```suggestion
        let committed_uuids = committed_uuids
            .into_iter()
            .map(|u| {
                CompressedRistretto(u)
                    .decompress()
                    .ok_or_else(||CodecError::from("Invalid UUID."))
            })
            .collect::<Result<Vec<_>,_>>()?;
```